python/lsst/analysis/tools/actions/keyedData/calcDistances.py

```diff
@@ -81,61 +89,139 @@ def __call__(self, data: KeyedData, **kwargs) -> KeyedData:
+            "separationResiduals": np.array([]),
+            "AMx": np.nan,
+            "ADx": np.nan,
+            "AFx": np.nan,
+        }
+
+        if len(data[self.groupKey]) == 0:
+            return distanceParams
+
+        rng = np.random.RandomState(seed=self.randomSeed)
+
+        def _compressArray(arrayIn):
+            h, rev = esutil.stat.histogram(arrayIn, rev=True)
+            arrayOut = np.zeros(len(arrayIn), dtype=np.int32)
+            (good,) = np.where(h > 0)
+            counter = 0
+            for ind in good:
+                arrayOut[rev[rev[ind] : rev[ind + 1]]] = counter
+                counter += 1
+            return arrayOut
+
+        groupId = _compressArray(data[self.groupKey])
+
+        nObj = groupId.max() + 1
+
+        # Compute the meanRa/meanDec.
+        meanRa = np.zeros(nObj)
+        meanDec = np.zeros_like(meanRa)
+        nObs = np.zeros_like(meanRa, dtype=np.int64)
+
+        # Check if tract is overlapping ra=0 and rotate if so.
+        # We assume a tract is smaller than 60x60 degrees.
+        rotation = 0.0
+        if np.max(data[self.raKey]) > 330.0 and np.min(data[self.raKey]) < 30.0:
+            rotation = 180.0
+            raRotated = np.array(data[self.raKey]) - rotation
+        else:
+            raRotated = np.array(data[self.raKey])
+
+        np.add.at(meanRa, groupId, raRotated)
+        np.add.at(meanDec, groupId, np.array(data[self.decKey]))
+        np.add.at(nObs, groupId, 1)
+
+        meanRa /= nObs
+        meanDec /= nObs
+        meanRa += rotation
+
+        D = (self.annulus * u.arcmin).to_value(u.degree)
+        width = (self.width * u.arcmin).to_value(u.degree)
+        annulus = D + (width / 2) * np.array([-1, +1])
+
+        # Match this catalog to itself within the radius and then cut
+        # to the annulus inner radius.
+        with Matcher(meanRa, meanDec) as m:
+            idx, i1, i2, d = m.query_self(annulus[1], return_indices=True)
+
+        inAnnulus = d > annulus[0]
+        i1 = i1[inAnnulus]
+        i2 = i2[inAnnulus]
+
+        if len(i1) == 0:
+            return distanceParams
+
+        if len(i1) > self.maxPairs:
+            # Downsample the pairs.
+            selection = rng.choice(len(i1), size=self.maxPairs, replace=False)
+            i1 = i1[selection]
+            i2 = i2[selection]
+
+        # Match groups and get indices.
+        h, rev = esutil.stat.histogram(groupId, rev=True)
+
+        # Match together pairs that have the same visit.
+        # It is unfortunate that this requires a loop, but it is not slow.
+        # After this matching we have a set of matchedObsInd1/matchedObsInd2
+        # that are all individual observations that are in the annulus and
+        # share a visit. The matchedPairInd groups all the paired observations
+        # of a given pair.
+        matchedObsInd1 = []
+        matchedObsInd2 = []
+        matchedPairInd = []
+        for ind in range(len(i1)):
```

Comment:
I didn't test this, but I think you could do this loop over visits instead of over matches, which would mean fewer iterations of the loop. Since you say it's not slow though, it's probably not worth reworking.

Reply:
You know, I thought that too, but couldn't figure out how to make it work with the matching necessary.  Although I do think an even more optimized version (maybe even without this loop) is possible, I had to just stop and accept the speed increase I had gotten without making it perfect.